src/serializer/Emitter.js

```diff
@@ -175,7 +198,29 @@ export class Emitter {
       this._residualFunctions.addFunctionUsage(val, this.getBodyReference());
       return undefined;
     } else if (val instanceof AbstractValue) {
-      if (val.hasIdentifier() && !this._declaredAbstractValues.has(val)) return val;
+      if (val.hasIdentifier()) {
+        const valSerializeBodyStack = this._declaredAbstractValues.get(val);
+        if (!valSerializeBodyStack) {
+          // Hasn't been serialized yet.
+          return val;
+        } else {
+          // The dependency has already been serialized(declared). But we may still have to wait for
+          // current generator body to be available, under following conditions:
+          // 1. Not delay initialization.
+          // 2. Not emitting in current active generator.(otherwise no need to wait)
+          // 3. Dependency's active ancestor generator body is lower in generator stack than current body.
+          const valActiveAncestorBody = this._getFirstAncestorGeneratorWithActiveBody(valSerializeBodyStack);
+          invariant(this._activeBodies.includes(valActiveAncestorBody));
+          if (!this._activeBodies.includes(this._body)) {
+            invariant(this._delayInitializations);
```

Comment:
My assumption/understanding is that: the only time this._activeBodies does not contain this._body happens during delay initialization scenario, which this._body points to the body of the residual function. In this scenario, since we are emitting into residual function body, we do not need to wait depdendencies as long as it is declared.

I can not come up a real world case that violates this assumption. Let me know if you can.